test_cluster_migration_with_tiering_and_deletes fails with off-by-one key count mismatch

The failure occurred on my PR (#6674) but is completely unrelated to the changes (my PR only touches `MGET`/`GAT` logic).

**CI Link:** [https://github.com/dragonflydb/dragonfly/actions/runs/22180622257/job/64140668595?pr=6674](https://github.com/dragonflydb/dragonfly/actions/runs/22180622257/job/64140668595?pr=6674)

**Failure Output:**
The test fails on the final assertion where the destination node's key count is off by exactly one key:

```python
>       assert info["db0"]["keys"] == keys - delete_succeded
E       assert equals failed
E         962938  962937

```
Right before the assertion fails, the replica throws an I/O error on the migration connection.

From `dragonfly.WARNING` (Process 22804):
```text
W20260219 12:06:25.729902 22804 execution_state.cc:93] ReportError: Input/output error

```
Logs attached.

[large-tests-arm-logs.zip](https://github.com/user-attachments/files/25415626/large-tests-arm-logs.zip)

--- a/src/server/journal/streamer.cc
+++ b/src/server/journal/streamer.cc
@@ -602,6 +602,11 @@ bool RestoreStreamer::WriteBucket(PrimeTable::bucket_iterator it, const ExpireTa
   // Only track tiered keys when needed and flush delayed entries
   // 1. When we have tiered storage
   // 2. We're called from a OnDbChange callback
+  //
+  // We need to track all keys in bucket with tiering. Even if they are not set as external. There
+  // is situation when we request externalization of key and key is read - marking it as not
+  // external but not yet flushed. When OnDbChange callback is called we need to flush it and than
+  // write journal changes - so we cannot realy on IsExternal flag and need to track all keys.
   const bool track_tiered_keys =
       on_db_change_cb && EngineShard::tlocal()->tiered_storage() != nullptr;
 
@@ -622,7 +627,7 @@ bool RestoreStreamer::WriteBucket(PrimeTable::bucket_iterator it, const ExpireTa
           expire = db_slice_->ExpireTime(eit->second);
         }
         // Track tiered keys that will need delayed entry flushing
-        if (track_tiered_keys && pv.IsExternal() && !pv.IsCool()) {
+        if (track_tiered_keys) {
           tiered_keys.emplace(key);
         }
         WriteEntry(key, it->first, pv, expire);
@@ -636,9 +641,8 @@ bool RestoreStreamer::WriteBucket(PrimeTable::bucket_iterator it, const ExpireTa
     // for force-flushing their delayed entries
     if (track_tiered_keys) {
       for (it.AdvanceIfNotOccupied(); !it.is_done(); ++it) {
-        const auto& pv = it->second;
         string_view key = it->first.GetSlice(&key_buffer);
-        if (ShouldWrite(key) && pv.IsExternal() && !pv.IsCool()) {
+        if (ShouldWrite(key)) {
           tiered_keys.emplace(key);
         }
       }

--- a/tests/dragonfly/cluster_test.py
+++ b/tests/dragonfly/cluster_test.py
@@ -3735,6 +3735,10 @@ async def test_cluster_migration_with_tiering_and_deletes(df_factory: DflyInstan
     keys = 1000000
     await nodes[0].client.execute_command(f"DEBUG POPULATE {keys} key 440")
 
+    # Expect that number of added keys is 1000000
+    info = await nodes[0].client.info("keyspace")
+    assert info["db0"]["keys"] == keys
+
     # Wait for some data to be offloaded to tiered storage
     await asyncio.sleep(10)
 

